File: src/frontend/src/services/dataLineage.js
// [AI GENERATED]
// Model: GitHub Copilot (GPT-4.1)
// Logic: Scaffold a frontend service for Data Lineage & Provenance API endpoints.
// Why: Enables the React dashboard to record, retrieve, and trace data lineage events.
// Root Cause: No frontend integration existed for the new backend API.
// Context: This service provides methods for posting lineage events, fetching all events, and tracing data origins. Future: add error handling, persistent storage, and graph visualization. For advanced lineage analytics, consider using a more advanced model (Claude Opus).

import axios from 'axios';

const API_BASE = '/api';

export const recordLineageEvent = async (event) => {
    try {
        const res = await axios.post(`${API_BASE}/data-lineage/events`, event);
        return res.data;
    } catch (error) {
        console.error('Error recording lineage event:', error);
        throw new Error('Failed to record lineage event. Please try again later.');
    }
};

export const getLineageEvents = async (dataset) => {
    // Added dataset parameter validation
    if (dataset && typeof dataset !== 'string') {
        throw new Error('Invalid dataset parameter. It must be a string.');
    }

    try {
        const url = dataset ? `${API_BASE}/data-lineage/events?dataset=${encodeURIComponent(dataset)}` : `${API_BASE}/data-lineage/events`;
        const res = await axios.get(url);
        return res.data;
    } catch (error) {
        console.error('Error fetching lineage events:', error);
        throw new Error('Failed to fetch lineage events. Please try again later.');
    }
};

export const traceDataOrigin = async (dataset) => {
    // Added validation for dataset existence
    if (!dataset) {
        throw new Error('Dataset parameter is required.');
    }

    try {
        const res = await axios.get(`${API_BASE}/data-lineage/trace?dataset=${encodeURIComponent(dataset)}`);
        return res.data;
    } catch (error) {
        console.error('Error tracing data origin:', error);
        throw new Error('Failed to trace data origin. Please try again later.');
    }
};

File: backend/core/routers/data_lineage.py

# [AI GENERATED]
# Model: GitHub Copilot (GPT-4.1)
# Logic: Scaffold a FastAPI router for Data Lineage & Provenance, providing endpoints to record, retrieve, and trace data transformations and origins.
# Why: Roadmap item 22 requires end-to-end traceability for compliance, debugging, and auditability.
# Root Cause: No API existed for tracking or querying data lineage and provenance.
# Context: This router provides endpoints for submitting lineage events, retrieving lineage for a dataset, and tracing data origins. Future: integrate with persistent storage, add graph traversal, and support visualization. For advanced lineage analytics, consider using a more advanced model (Claude Opus).

from fastapi import APIRouter, status
from pydantic import BaseModel
from typing import List, Optional
from datetime import datetime

router = APIRouter()

# In-memory store for demo purposes
LINEAGE_EVENTS = []

class LineageEvent(BaseModel):
    id: int
    timestamp: datetime
    dataset: str
    operation: str  # e.g., 'ingest', 'transform', 'join', 'export'
    source_datasets: Optional[List[str]] = None
    user: Optional[str] = None
    details: Optional[str] = None

@router.post("/data-lineage/events", status_code=status.HTTP_201_CREATED)
def record_lineage_event(event: LineageEvent):
    LINEAGE_EVENTS.append(event)
    return {"message": "Lineage event recorded", "event_id": event.id}

@router.get("/data-lineage/events", response_model=List[LineageEvent])
def get_lineage_events(dataset: Optional[str] = None):
    if dataset:
        return [e for e in LINEAGE_EVENTS if e.dataset == dataset]
    return LINEAGE_EVENTS

@router.get("/data-lineage/trace", response_model=List[LineageEvent])
def trace_data_origin(dataset: str):
    # Simple trace: return all events where this dataset is a source or target
    return [e for e in LINEAGE_EVENTS if e.dataset == dataset or (e.source_datasets and dataset in e.source_datasets)]
